src/rpm/headers/types.rs

```diff
@@ -68,108 +70,81 @@ impl From<u16> for FileMode {
         // <https://en.wikipedia.org/wiki/Bitwise_operation#AND>
         let file_type = raw_mode & FILE_TYPE_BIT_MASK;
         let permissions = raw_mode & PERMISSIONS_BIT_MASK;
-        match file_type {
-            DIR_FILE_TYPE => FileMode::Dir { permissions },
-            REGULAR_FILE_TYPE => FileMode::Regular { permissions },
-            SYMBOLIC_LINK_FILE_TYPE => FileMode::SymbolicLink { permissions },
-            _ => FileMode::Invalid {
+        let file_type = match file_type {
+            DIR_FILE_TYPE => Ok(FileType::Dir),
+            REGULAR_FILE_TYPE => Ok(FileType::Regular),
+            SYMBOLIC_LINK_FILE_TYPE => Ok(FileType::SymbolicLink),
+            _ => Err(Self::Error::InvalidFileMode {
                 raw_mode: raw_mode as i32,
                 reason: "unknown file type",
-            },
-        }
+            }),
+        }?;
+
+        Ok(FileMode {
+            file_type,
+            permissions,
+        })
     }
 }
 
-impl From<i32> for FileMode {
-    fn from(raw_mode: i32) -> Self {
+impl TryFrom<i32> for FileMode {
+    type Error = errors::Error;
+
+    fn try_from(raw_mode: i32) -> Result<Self, Self::Error> {
         // since we ultimately only deal with 16bit integers
         // we need to check if a safe conversion to 16bit is doable.
         if raw_mode > u16::MAX.into() || raw_mode < i16::MIN.into() {
-            FileMode::Invalid {
+            Err(Self::Error::InvalidFileMode {
                 raw_mode,
                 reason: "provided integer is out of 16bit bounds",
-            }
+            })
         } else {
-            FileMode::from(raw_mode as u16)
+            Ok(FileMode::try_from(raw_mode as u16)?)
         }
     }
 }
 
+// @todo: 0o7777? should it be 0o777?
```

Comment:
Apparently this is legit

> 
>  Let's fix up cdplayer with a liberal sprinkling of %attrs. Here's what the %files list looks like after we've had our way with it:
> 
>> %files
>> %attr(-, root, root) %doc README
>> %attr(4755, root, root) /usr/local/bin/cdp
>> %attr(-, root, root) /usr/local/bin/cdplay
>> %attr(-, root, rot) /usr/local/man/man1/cdp.1
>           
> 
> A couple points are worth noting here. The line for README shows that multiple macros can be used on a line — in this case, one to set file attributes, and one to mark the file as being documentation. The %attr for /usr/local/bin/cdp declares the file to be setuid root. If it sends a shiver down your spine to know that anybody can create a package that will run setuid root when installed on your system — Good! Just because RPM makes it easy to install software doesn't mean that you should blindly install every package you find.
> 
> A single RPM command can quickly point out areas of potential problems and should be issued on any package file whose creators you don't trust:
> 
>> % rpm -qlvp ../RPMS/i386/cdplayer-1.0-1.i386.rpm
>> drwxr-xr-x-  root  root   1024 Sep 13 20:16 /usr/doc/cdplayer-1.0-1
>> -rw-r--r---  root  root   1085 Nov 10 01:10 /usr/doc/cdplayer-1.0-1/README
>> -rwsr-xr-x-  root  root  40739 Sep 13 21:32 /usr/local/bin/cdp
>> lrwxrwxrwx-  root  root     47 Sep 13 21:32 /usr/local/bin/cdplay -> ./cdp
>> -rwxr-xr-x-  root   rot   4550 Sep 13 21:32 /usr/local/man/man1/cdp.1
>> % 

This almost makes me want to push the setuid/setgid bits into their own methods, but that would be a real pain for anyone trying to implement something that builds from spec files.